File: JUNHEE/프로그래머스_week11/땅따먹기.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int solution(vector<vector<int> > land)
{
    int answer = 0;
    vector<vector<int>> map(land.size(), vector<int>(4));
    for(int i = 0; i < 4; i++){
        map[0][i] = land[0][i];
    }
    for(int i = 1; i < land.size(); i++){
        for(int j = 0; j < 4; j++){
            int temp = land[i][j];
            for(int k = 0; k < 4; k++){
                if(j != k){
                    if(temp + map[i-1][k] > map[i][j])
                        map[i][j] = map[i-1][k] + temp;
                }
            }
        }
    }
    int last = map.size() - 1;
    answer = max({map[last][0], map[last][1], map[last][2], map[last][3]});
    return answer;
}

File: DONGUK/프로그래머스_WEEK_11/땅 따먹기.py
def solution(land):
    answer = 0

    dp = [[0] * 4 for _ in range(len(land))]

    for i in range(4):
        dp[0][i] = land[0][i]

    N = len(land)
    for i in range(1, N):
        for j in range(4):
            data = []
            for k in range(4):
                if k != j:
                    data.append(dp[i-1][k])

            dp[i][j] = max(data) + land[i][j]
            
    answer = max(dp[-1])
    return answer